### ISSUE TITLE
Problem with singleton unregister in singleton register chapter

### ISSUE THREAD

**Issue by StatisMike** [godot-rust/book#30]
Problem with current _singleton unregistering_ method at https://godot-rust.github.io/book/recipes/engine-singleton.html#registering-a-singleton

```rust
    fn on_level_deinit(level: InitLevel) {
        if level == InitLevel::Scene {
            // Unregistering is needed to avoid memory leaks and 
            // warnings, especially for hot reloading.
            Engine::singleton().unregister_singleton(
                StringName::from("MyEditorSingleton")
            );
        }
    }
```

The thing I've observed during my playing with singletons, the `unregister_singleton` call doesn't actually free the object. Correct implementation to avoid memory leaks would be a little more verbose:

```rust
    fn on_level_deinit(level: InitLevel) {
        if level == InitLevel::Scene {
            // Retrieve the object to free it after unregistering
            let mut my_editor_singleton = Engine.singleton().get_singleton(StringName::from("MyEditorSingleton"));

            // Unregister the singleton from Godot's singletons list
            Engine::singleton().unregister_singleton(
                StringName::from("MyEditorSingleton")
            );

            // Free the object itself
           my_editor_singleton.free();
        }
    }
```

The question stands if this isn't possibly linked with [RefCounted singletons issue](https://github.com/godot-rust/gdext/issues/522) from `gdext` and will be fixed later or is the intended behaviour on upstream. The GDScript `unregister_singleton()` doesn't leave dangling objects, but its possible that `GDScript` is expected to just clean them automatically, making the problem on `GDExtension`s side.

Based on the question answer, the solutions would vary (maybe some other pattern would be preferred). The above works in current state of `gdext`. If its imperative to have this corrected for now, I could provide PR to book as soon as I figure out how built it correctly on local :)

**Comment by Bromeon** [comment]
> The question stands if this isn't possibly linked with [RefCounted singletons issue](https://github.com/godot-rust/gdext/issues/522) from `gdext` and will be fixed later or is the intended behaviour on upstream.

I don't think so, since the user class explicitly inherits `Object`, not `RefCounted`.

You're right, registering/unregistering singletons is independent of creation/destruction, so this was an oversight. Eventually I plan to have a more ergonomic API for singletons, but until then we should fix this 🙂

**Comment by StatisMike** [comment]
Godot engine may expect the GDExtension to handle the object itself on its side independently from itself. Could be beneficial for the future singleton API to keep the `gdext`-registered singletons in some kind of collection internally, which could be then automatically freed on level deinit. From my current grasp on the topic, it could potentially also fix the RefCounted issue.

On that note - I will also mention the workaround for RefCounted singletons in my PR, as it fixed some issues for me.

**Comment by StatisMike** [comment]
A #31 was merged, this issue can be safely closed.

**Comment by Bromeon** [comment]
For the future, you can directly write

> Closes #30.

in the initial message of the PR, this will automate the closing 🙂
